ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java

```diff
@@ -438,4 +532,41 @@ public FunctionRegistry getFunctionRegistry() {
   public SchemaRegistryClient getSchemaRegistryClient() {
     return schemaRegistryClient;
   }
+
+  protected static class KsqlValueJoiner
+      implements ValueJoiner<GenericRow, GenericRow, GenericRow> {
+    private final Schema leftSchema;
+    private final Schema rightSchema;
+
+    KsqlValueJoiner(final Schema leftSchema, final Schema
+        rightSchema) {
+      this.leftSchema = leftSchema;
+      this.rightSchema = rightSchema;
+    }
+
+    @Override
+    public GenericRow apply(final GenericRow left, final GenericRow right) {
+      final List<Object> columns = new ArrayList<>();
+      if (left != null) {
+        columns.addAll(left.getColumns());
+      } else {
+        fillWithNulls(columns, leftSchema.fields().size());
+      }
+
+      if (right != null) {
+        columns.addAll(right.getColumns());
+      } else {
+        fillWithNulls(columns, rightSchema.fields().size());
+      }
+
+      return new GenericRow(columns);
+    }
+
+    private void fillWithNulls(final List<Object> columns, final int numToFill) {
+      for (int i = 0; i < numToFill; ++i) {
```

Comment:
Consider making this more functional by returning a new list instead of updating: ```return Arrays.asList(new Object[numToFill]);```